### PR TITLE
[core][adag] Separate the outputs of execute and execute_async to multiple refs or futures to allow clients to retrieve them one at a time (#46908)

### LINKED ISSUE #46908: [core][adag] aDAGs with multiple outputs should allow getting them one at a time
### Description

Right now, if you use MultiOutputNode to wrap a DAG's outputs, you get back a single CompiledDAGRef that points to a list of all of the outputs. That means that even if you only need one of the outputs, you have to get and deserialize all of them at the same time. This API is also a bit strange compared to Ray's usual API, which lets you get outputs individually if a task has multiple returns.

We should update the aDAG API to return one CompiledDAGRef per output, something like this:
```python
with InputNode() as inp:
  dag = MultiOutputNode([a.foo.bind(inp), b.foo.bind(inp)])

dag = dag.experimental_compile()
# Now returns two refs, one per output in MultiOutputNode.
ref1, ref2 = dag.execute(1)
```

### Use case

One use case is in vLLM tensor parallelism. Since all shards return the same results, we only need to get results from one of the workers.

### PR DESCRIPTION
**Pull request by jeffreyjeffreywang**
## Why are these changes needed?
Currently, if `MultiOutputNode` is used to wrap a DAG's output, you get back a single `CompiledDAGRef` or `CompiledDAGFuture`, depending on whether `execute` or `execute_async` is invoked, that points to a list of all of the outputs. To retrieve one of the outputs, you have to get and deserialize all of them at the same time.

This PR separates the output of `execute` and `execute_async` to a list of `CompiledDAGRef` or `CompiledDAGFuture` when the output is wrapped by `MultiOutputNode`. This is particularly useful for vLLM tensor parallelism. Since all shards return the same results, we only need to fetch result from one of the workers.

## Related issue number

Resolves #46908 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by jeffreyjeffreywang** [comment]
Hi Ray developers, I would love to have some help validating `test_execution_schedule_gpu.py` tests as they require a GPU device which I don't have one.

File: python/ray/dag/compiled_dag_node.py
Comment (ruisearch42) [inline python/ray/dag/compiled_dag_node.py:1595]: `timeout` needs to be updated in this loop?

File: python/ray/dag/compiled_dag_node.py
Comment (ruisearch42) [inline python/ray/dag/compiled_dag_node.py:1591]: Do we need to fetch from all channels? Can you please add documentation?

**Review by ruisearch42 — COMMENTED** [review]
Fist pass. Looks pretty nice. Thanks for the contribution!

File: python/ray/dag/compiled_dag_node.py
Comment (jeffreyjeffreywang) [reply python/ray/dag/compiled_dag_node.py:1591]: Yes, this aligns with the original behavior. The original code has a single fetcher which encapsulates all output channels, and invoking `read` on that fetcher (`_dag_output_fetcher`) fetches results from all output channels. Now, we have a fetcher for each channel (`_dag_output_fetchers`), and each fetcher is only responsible for getting results for its channel. The reason for this refactor is that we will now be able to get result from a single output channel so that we are not forced to get all results from all channels. I'll add documentation to make it clearer.

File: python/ray/dag/compiled_dag_node.py
Comment (jeffreyjeffreywang) [reply python/ray/dag/compiled_dag_node.py:1595]: Good catch, yeah, reading from each channel should deduct from the global timeout.

File: python/ray/dag/compiled_dag_node.py
Comment (rkooo567) [inline python/ray/dag/compiled_dag_node.py:633]:
```suggestion
        # execution_index -> {channel -> result}
        self._result_buffers: Dict[int, Dict[ChannelInterface, Any]] = {}
```

File: python/ray/dag/compiled_dag_node.py
Comment (rkooo567) [inline python/ray/dag/compiled_dag_node.py:1553]: can you update a docstring?

File: python/ray/dag/compiled_dag_node.py
Comment (rkooo567) [inline python/ray/dag/compiled_dag_node.py:1573]: why do we need this if now?

File: python/ray/dag/compiled_dag_node.py
Comment (rkooo567) [inline python/ray/dag/compiled_dag_node.py:1581]: I think this is wrong now. We should do len(self._result_buffer) * num_output_channels

Comment (rkooo567) [reply python/ray/dag/compiled_dag_node.py:1581]: can you also add a unit test? (or modify existing test)

File: python/ray/dag/compiled_dag_node.py
Comment (rkooo567) [inline python/ray/dag/compiled_dag_node.py:1594]: can you just use

```
self._result_buffer = defaultdict(dict)
```

and we can remove this code

File: python/ray/dag/compiled_dag_node.py
Comment (rkooo567) [inline python/ray/dag/compiled_dag_node.py:1596]: QQ @ruisearch42 do you know if there is any regression by doing this? (now we are blocking read each channel as many as output ref times)

File: python/ray/dag/compiled_dag_node.py
Comment (rkooo567) [inline python/ray/dag/compiled_dag_node.py:1714]: QQ: can we use asyncio.gather instead and do all put together?

**Review by rkooo567 — COMMENTED** [review]
I think generally the approach makes sense! But I am seeing 2 problems.

1. It is backward incompatible changes and will break vllm. How complicated is it to support `MultiOutputNode(_multiple_return_refs=True)`? I am not suggesting to do it, but I'd like to measure the complexity 
2. I think although we don't call ray.get, ray.get is called and deserialization still happens when the dag ref is deallocated because of this code https://github.com/ray-project/ray/blob/4aea49f5301ac083aed5d016919482afd2f54f83/python/ray/experimental/compiled_dag_ref.py#L80. I think we should improve this part of code to not doing deserialization. I think this part is a little bit tricky to handle (so we can probably handle as a follow up)

File: python/ray/dag/compiled_dag_node.py
Comment (jeffreyjeffreywang) [reply python/ray/dag/compiled_dag_node.py:1594]: Sure thing, will do!

**Comment by jeffreyjeffreywang** [comment]
Hey @rkooo567, thank you so much for the review. Let me experiment a few things you suggested and get back to you tomorrow.

**Review by stephanie-wang — COMMENTED** [review]
Thanks for this contribution!

File: python/ray/dag/compiled_dag_node.py
Comment (jeffreyjeffreywang) [reply python/ray/dag/compiled_dag_node.py:1573]: An ImportError will occur with the original code when destructing unused CompiledDAGRef (ones that we have not called get()). Here is a minimal repro for the original behavior:

```
foo = Foo.remote()
bar = Bar.remote()

with InputNode() as inp:
    dag = MultiOutputNode([foo.increment.bind(inp), bar.decrement.bind(inp)])

dag = dag.experimental_compile()

ref1 = dag.execute(1)
ref2 = dag.execute(1)

assert ref2.get() == [2, -2]

dag.teardown()
# When exiting the program, ref1.__del__ is invoked.
# Since it has not been called with get(), ref1.get() will
# be invoked subsequently. In _execute_until, even though
# the DAG won't be executed again, we still attempt to
# import a library, namely ray.dag.DAGContext. Since
# Python is shutting down, an ImportError occurs
# (ImportError: sys.meta_path is None, Python is likely
# shutting down).
```

The same problem persists with the current behavior if the if clause is not introduced. Another the benefit of introducing the if clause is to only import the library and calculate `timeout` when necessary. `timeout` is only used when `self._max_execution_index < execution_index` is `True`. We shouldn't import the library when `self._max_execution_index == execution_index`. I should keep the following while loop out of the if clause though.

Comment (jeffreyjeffreywang) [reply python/ray/dag/compiled_dag_node.py:1573]: As I speak, I found out that if the last `CompiledDAGRef` is unused, the `ImportError` persists even when the if clause is introduced. Is there a guidance for Ray users to always invoke `get()` on all `CompiledDAGRef`? If not, this may be a bug that needs to be addressed in a separate PR. WDYT?

Here is a minimal repro:
```
foo = Foo.remote()
bar = Bar.remote()

with InputNode() as inp:
    dag = MultiOutputNode([foo.increment.bind(inp), bar.decrement.bind(inp)])

dag = dag.experimental_compile()

ref1 = dag.execute(1)
ref2 = dag.execute(1)

assert ref1.get() == [1, -1]

dag.teardown()
# Upon destruction, the DAG will be executed until the latest index. Again,
# we attempt to import DAGContext during program exit. 
```

Comment (jeffreyjeffreywang) [reply python/ray/dag/compiled_dag_node.py:1573]: For now, I think it's worth keeping the if clause and have another bug tracking the dangling `CompiledDAGRef` issue.

File: python/ray/dag/compiled_dag_node.py
Comment (jeffreyjeffreywang) [reply python/ray/dag/compiled_dag_node.py:1581]: Agree. `len(self._result_buffer)` indicates the number of DAG executions while `len(self._result_buffer) * num_output_channels` represents the total number of outputs. I will adjust the naming of `_max_buffered_results` to `_max_buffered_executions` (any other suggestions?) and add a unit test.

File: python/ray/dag/compiled_dag_node.py
Comment (jeffreyjeffreywang) [reply python/ray/dag/compiled_dag_node.py:1573]: I think this issue I mentioned is correlated with your second point. Is there any reason why we want to avoid execution result leak (which is why we currently execute the DAG until the latest index and `get()` all results)?

File: python/ray/dag/tests/experimental/test_accelerated_dag.py
Comment (jeffreyjeffreywang) [reply python/ray/dag/tests/experimental/test_accelerated_dag.py:232]: It seems like `ray.get` only supports `CompiledDAGRef`, `ObjectRef`, and a list of `ObjectRef`, and `CompiledDAGRef` is not a subclass of `ObejctRef` although it's declared to be so in its docstring. I'm not sure how `ray.get(refs)` will work. Am I missing anything?

File: python/ray/dag/compiled_dag_node.py
Comment (jeffreyjeffreywang) [reply python/ray/dag/compiled_dag_node.py:1654]: Great point, I'll adjust the logic.

File: python/ray/dag/compiled_dag_node.py
Comment (jeffreyjeffreywang) [reply python/ray/dag/compiled_dag_node.py:566]: Note: `self._result_buffer` and perhaps `self._dag_output_fetchers` need refactoring.

**Comment by jeffreyjeffreywang** [comment]
Thanks for all of the feedback. I'll keep addressing comments tomorrow and over the weekend. Will ping you as soon as the revision is ready for review.

**Comment by jeffreyjeffreywang** [comment]
> It is backward incompatible changes and will break vllm. How complicated is it to support MultiOutputNode(_multiple_return_refs=True)? I am not suggesting to do it, but I'd like to measure the complexity.

@rkooo567, I think this should be quite straightforward to support. Essentially, we can ask clients to specify `multiple_return_refs=True` when they want a list of refs rather than a single ref wrapping all outputs. `execute` and `execute_async` can check whether `MultiOutputNode` is used to wrap the output and whether `multiple_return_refs` is set to True. If both conditions are satisfied, we return a list of refs. Otherwise, fall back to original behavior. I'll get the change in for you to review.

File: python/ray/dag/compiled_dag_node.py
Comment (jeffreyjeffreywang) [reply python/ray/dag/compiled_dag_node.py:566]: `self._fut_queues` and `self._dag_output_fetchers` are changed to lists with indexing consistent with `self.dag_output_channels`. `self._result_buffer` is still a dict but no longer uses `ChannelInterface` for indexing (instead, `{execution_index->{channel_index->result}}`). The reason I kept it as a dict is that when we get and remove results from the buffer, the remaining results won't be out of order.

An alternative to use list for `self._result_buffer` is to mark the popped results as `None` instead of remove them from the buffer. However, the downside is that the size of the list remains the same even if we get all results from the buffer. From debugging perspective, I don't see a clear advantage for one over the other. Let me know if you prefer `self._result_buffer` to be a list as well.

File: python/ray/dag/compiled_dag_node.py
Comment (jeffreyjeffreywang) [reply python/ray/dag/compiled_dag_node.py]: No longer applicable as `self._dag_output_fetchers` is gotten rid off entirely. There is no longer a fetcher for each channel. Results from all channels for each execution index are fetched at the same time and cached in `_result_buffer`.

File: python/ray/dag/compiled_dag_node.py
Comment (jeffreyjeffreywang) [reply python/ray/dag/compiled_dag_node.py:1714]: No longer applicable as `dag_output_fetchers` is gotten rid of. Per @stephanie-wang's comment, for each execution index, I fetch results from all channels and cache them in `_result_buffer` rather than fetching results separately for each channel with their individual fetcher (which was what `_dag_output_fetchers` represent).

**Comment by jeffreyjeffreywang** [comment]
Please hold off reviewing as I'm now working on making it backward compatible. Thanks!

**Comment by jeffreyjeffreywang** [comment]
The change is now backward compatible. Working on unit tests and addressing the remaining comments.

File: python/ray/dag/compiled_dag_node.py
Comment (jeffreyjeffreywang) [reply python/ray/dag/compiled_dag_node.py]: Actually, I think we need to change `max_buffered_results` to `max_buffered_executions` if we were to change `_max_buffered_results` to `_max_buffered_executions`. Since there are quite a few references to `max_buffered_results` in the repo (including developer APIs and `python/ray/dag/context.py` which doesn't have much relevance to the changes in this PR), I'd suggest to not modify the naming now. We can track it as a bug and address it in a follow-up.

As for the unit test, I'll add one tomorrow.

**Comment by jeffreyjeffreywang** [comment]
All comments are addressed except for a test for validating an error is raised when the number of buffered results exceeds `_max_buffered_results` which I will complete tomorrow. @rkooo567 @stephanie-wang The rest of the PR is ready for another round of review. Thank you 🙂

File: python/ray/dag/compiled_dag_node.py
Comment (jeffreyjeffreywang) [reply python/ray/dag/compiled_dag_node.py]: Done adding a unit test for cases when there are multiple outputs. Now, if not all results from an execution index are fetched with `get()`, that execution index will still count towards the number of buffered results. Let me know if this sounds okay to you!

**Comment by jeffreyjeffreywang** [comment]
Done with adding a new unit test for validating max buffered result logic. @stephanie-wang @rkooo567 please let me know if you have any other concerns with the latest changes. Thanks!

**Comment by stephanie-wang** [comment]
> Done with adding a new unit test for validating max buffered result logic. @stephanie-wang @rkooo567 please let me know if you have any other concerns with the latest changes. Thanks!

Sorry for the delay, taking a look today!

**Comment by rkooo567** [comment]
yep will finish review by today as well!!

File: python/ray/dag/compiled_dag_node.py
Comment (stephanie-wang) [inline python/ray/dag/compiled_dag_node.py:1932]: I would split this into a cache setter and getter method to make this logic more robust, instead of allowing `result` to be optional.

Ideally, `_execute_until` should also use these methods.

File: python/ray/dag/compiled_dag_node.py
Comment (stephanie-wang) [inline python/ray/dag/compiled_dag_node.py:1956]: Doesn't this need to set `result`? Please add a test if not tested already.

File: python/ray/experimental/compiled_dag_ref.py
Comment (stephanie-wang) [inline python/ray/experimental/compiled_dag_ref.py:169]: Nice change, this looks pretty clean!

One nit: for readability, change the logic to:

```
if not cached:
  yield from fut
  cache values
pop from cache
```

**Review by stephanie-wang — COMMENTED** [review]
Overall structure looks good! Just some comments to unify the sync/async codepaths more, and ideally can we remove `self._multiple_return_refs`?

File: python/ray/dag/compiled_dag_node.py
Comment (jeffreyjeffreywang) [reply python/ray/dag/compiled_dag_node.py:705]: Will do!

File: python/ray/dag/compiled_dag_node.py
Comment (jeffreyjeffreywang) [reply python/ray/dag/compiled_dag_node.py:1819]: Yup no problem. I agree with being overly cautious although this path typically doesn't induce errors.

File: python/ray/dag/compiled_dag_node.py
Comment (jeffreyjeffreywang) [reply python/ray/dag/compiled_dag_node.py:1956]: Yes, this should. This could potentially return incorrect results when CompiledDAGFutures are fetched in order. Will add a test as well.

File: python/ray/dag/compiled_dag_node.py
Comment (jeffreyjeffreywang) [reply python/ray/dag/compiled_dag_node.py:705]: No longer applicable as this attribute is removed from CompiledDAG. I'm now only using `_has_single_output` as the flag to determine whether we want to return a list of refs/futures. Adjusted the property `multiple_return_refs` to `return_multiple_refs` for `MultiOutputNode` though.

File: python/ray/dag/compiled_dag_node.py
Comment (jeffreyjeffreywang) [reply python/ray/dag/compiled_dag_node.py:1882]: Yep, removed `multiple_return_refs`.

**Comment by jeffreyjeffreywang** [comment]
> Overall structure looks good! Just some comments to unify the sync/async codepaths more, and ideally can we remove self._multiple_return_refs?

Thanks for the feedback, Stephanie! I refactored the code, unified sync/async codepaths, and added one more test case. I think we still need `multiple_return_refs` though (see my comments in the relevant thread).

Here are some pending issues that we might want to address as follow-ups according to our discussion above:

- [ ] Deserialization still happens in the destructor of CompiledDAGRef/CompiledDAGFuture
- [ ] ray.get doesn't support fetching results from a list of CompiledDAGRef
- [ ] There isn't sufficient ADAG test coverage for async cases (quite comprehensive for sync cases though)

File: python/ray/dag/compiled_dag_node.py
Comment (stephanie-wang) [reply python/ray/dag/compiled_dag_node.py]: Hmm sorry I didn't quite understand this. I think the two cases are:

DAG contains a MultiOutputNode -> `_has_single_output=True` (and `returns_multiple_refs=True`)
DAG does not contain a MultiOutputNode -> `_has_single_output=False` (and `returns_multiple_refs=False`)

Can you explain which case is the one where `_has_single_output != returns_multiple_refs`?

**Comment by stephanie-wang** [comment]
This looks great!

Two last things before we merge:
- could you answer my question in the other thread about why we need returns_multiple_refs?
- "ray.get doesn't support fetching results from a list of CompiledDAGRef" -> can we support this? Otherwise we will have a breaking API change. A naive solution that just `gets` the refs in order is good for now.

File: python/ray/dag/compiled_dag_node.py
Comment (jeffreyjeffreywang) [reply python/ray/dag/compiled_dag_node.py]: - DAG contains a MultiOutputNode `node`
    - `node.returns_multiple_refs=True`: `has_single_output=True`, `returns_multiple_refs=True`
    - `node.returns_multiple_refs=False`: `has_single_output=False`, `returns_multiple_refs=False`
- DAG does not contain a MultiOutputNode
    - `has_single_output=True`, `returns_multiple_refs=False`

Here is a minimal repro:
```
a = Actor.remote(0)
with InputNode() as i:
    dag = a.echo.bind(i)

compiled_dag = dag.experimental_compile()
print(compiled_dag._has_single_output) # True
print(compiled_dag._returns_multiple_refs) # False
```

Let me know if this makes sense to you 🙂

**Comment by jeffreyjeffreywang** [comment]
> "ray.get doesn't support fetching results from a list of CompiledDAGRef" -> can we support this? Otherwise we will have a breaking API change. A naive solution that just gets the refs in order is good for now.
Definitely, will do 😃

**Comment by jeffreyjeffreywang** [comment]
Done supporting fetching a list of CompiledDAGRef with `ray.get()`. @stephanie-wang, let me know if you have further concerns. Thank you again for all of your feedback!

File: python/ray/dag/compiled_dag_node.py
Comment (stephanie-wang) [reply python/ray/dag/compiled_dag_node.py]: I would simplify it to these cases:
- DAG contains a MultiOutputNode (which may only have one ref): `has_single_output=False`
- DAG does not contain a MultiOutputNode: `has_single_output=True`

Basically the first case should always return a list of refs, even if there is only one ref. You could consider renaming it to something like `returns_list`, to make it clearer.

File: python/ray/_private/worker.py
Comment (stephanie-wang) [inline python/ray/_private/worker.py:2666]:
```suggestion
            all_compiled_dag_refs = True
```
(we use underscore naming for python)

File: python/ray/_private/worker.py
Comment (stephanie-wang) [inline python/ray/_private/worker.py:2672]: I think we also need to handle the else condition. It's okay to throw an error for now if the list contains a CompiledDAGRef but not all CompiledDAGRefs.

File: python/ray/dag/compiled_dag_node.py
Comment (jeffreyjeffreywang) [reply python/ray/dag/compiled_dag_node.py]: I introduced `returns_multiple_refs` as an attribute for MultiOutputNode per my discussion with SangBin.

> It is backward incompatible changes and will break vllm. How complicated is it to support MultiOutputNode(_multiple_return_refs=True)? I am not suggesting to do it, but I'd like to measure the complexity.

`execute` or `execute_async` now return a single ref/future even if there is a MultiOutputNode as long as the node's `returns_multiple_refs` attribute is set to False (which is by default). This prevents from breaking existing consumer code as they expect to get a single ref/future from `execute` and `execute_async`. Only new consumer code which specifies `returns_multiple_refs=True` will get a list of ref/future when there is a MultiOutputNode. Do you think this makes sense?

Comment (stephanie-wang) [reply python/ray/dag/compiled_dag_node.py]: Thanks for the explanation! I synced with @rkooo567 offline and we agreed that it should be okay to go ahead with this change and remove `_multiple_return_refs` (aDAGs are still in alpha so for now breaking changes are okay). @rkooo567 can shepherd the vLLM change required.

Comment (rkooo567) [reply python/ray/dag/compiled_dag_node.py]: sorry @jeffreyjeffreywang you have to be back and fourth here :(.

Comment (jeffreyjeffreywang) [reply python/ray/dag/compiled_dag_node.py]: @rkooo567 No worries! I haven't taken the fact that aDAG being alpha into account either. @stephanie-wang thank you for the update, I'll revert the changes related to `_multiple_return_refs` later today.

File: python/ray/dag/compiled_dag_node.py
Comment (jeffreyjeffreywang) [inline python/ray/dag/compiled_dag_node.py:1450]: Revert

File: python/ray/dag/compiled_dag_node.py
Comment (jeffreyjeffreywang) [reply python/ray/dag/compiled_dag_node.py]: Done reverting :)

**Review by stephanie-wang — APPROVED** [review]
Awesome, thank you for this great contribution!

By the way, if you are not already on it, please feel free to join #ray-accelerated-dags at ray.slack.com. I'm sure there are some interested users there :)

**Comment by jeffreyjeffreywang** [comment]
Thank you so much @stephanie-wang for all of the reviews. I'm looking forward to contributing more 😄 

Would you mind taking a final look at my latest update? I resolved some merge conflicts with SangBin's recent changes.

**Comment by jeffreyjeffreywang** [comment]
> I think although we don't call ray.get, ray.get is called and deserialization still happens when the dag ref is deallocated because of this code. https://github.com/ray-project/ray/blob/4aea49f5301ac083aed5d016919482afd2f54f83/python/ray/experimental/compiled_dag_ref.py#L80
> I think we should improve this part of code to not doing deserialization. I think this part is a little bit tricky to handle (so we can probably handle as a follow up)

Tracking this issue with #47614.

**Comment by stephanie-wang** [comment]
> Thank you so much @stephanie-wang for all of the reviews. I'm looking forward to contributing more. 😄
> 
> Would you mind taking a final look at my latest update? I resolved some merge conflicts with SangBin's recent changes.

It looks good to me! Will merge once tests pass.

**Comment by rkooo567** [comment]
also to be clear @stephanie-wang it will be a breaking change right?

**Comment by jeffreyjeffreywang** [comment]
> also to be clear @stephanie-wang. @jeffreyjeffreywang it will be a breaking change right?

Correct, this will be breaking. `execute` and `execute_async` on adag with `MultiOutputNode` will always return a list of refs/futures.

**Comment by rkooo567** [comment]
makes sense! Also @jeffreyjeffreywang do we have issues for all follow up problems?

```
- Deserialization still happens in the destructor of CompiledDAGRef/CompiledDAGFuture
-  ray.get doesn't support fetching results from a list of CompiledDAGRef
-  There isn't sufficient ADAG test coverage for async cases (quite comprehensive for sync cases though)
```

**Comment by jeffreyjeffreywang** [comment]
> makes sense! Also @jeffreyjeffreywang do we have issues for all follow up problems?

Yup, the second one is addressed in this PR. #47662 and #47614 track the remaining follow-up problems.

**Comment by rkooo567** [comment]
@jeffreyjeffreywang https://github.com/ray-project/ray/issues/47684

actuallay is this a known issue? can you take a look?

**Comment by jeffreyjeffreywang** [comment]
@rkooo567 Yeah, I also realized this may be a problem over the weekend. I think we need to support calling `ray.wait` on `CompiledDAGFuture` and a list of `CompiledDAGFuture` as well.

**Pull request merged**
